parquet-variant/src/main/java/org/apache/parquet/variant/VariantUtil.java

```diff
@@ -0,0 +1,843 @@
+  }
+
+  /**
+   * Returns the value type of Variant value `value[pos...]`. It is only legal to call `get*` if
+   * `getType` returns the corresponding type. For example, it is only legal to call
+   * `getLong` if this method returns `Type.Long`.
+   * @param value The Variant value to get the type from
+   * @param pos The starting index of the Variant value
+   * @return The type of the Variant value
+   */
+  public static Type getType(byte[] value, int pos) {
+    checkIndex(pos, value.length);
+    int basicType = value[pos] & BASIC_TYPE_MASK;
+    int typeInfo = (value[pos] >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    switch (basicType) {
+      case SHORT_STR:
+        return Type.STRING;
+      case OBJECT:
+        return Type.OBJECT;
+      case ARRAY:
+        return Type.ARRAY;
+      default:
+        switch (typeInfo) {
+          case NULL:
+            return Type.NULL;
+          case TRUE:
+          case FALSE:
+            return Type.BOOLEAN;
+          case INT8:
+            return Type.BYTE;
+          case INT16:
+            return Type.SHORT;
+          case INT32:
+            return Type.INT;
+          case INT64:
+            return Type.LONG;
+          case DOUBLE:
+            return Type.DOUBLE;
+          case DECIMAL4:
+          case DECIMAL8:
+          case DECIMAL16:
+            return Type.DECIMAL;
+          case DATE:
+            return Type.DATE;
+          case TIMESTAMP:
+            return Type.TIMESTAMP;
+          case TIMESTAMP_NTZ:
+            return Type.TIMESTAMP_NTZ;
+          case FLOAT:
+            return Type.FLOAT;
+          case BINARY:
+            return Type.BINARY;
+          case LONG_STR:
+            return Type.STRING;
+          case TIME:
+            return Type.TIME;
+          case TIMESTAMP_NANOS:
+            return Type.TIMESTAMP_NANOS;
+          case TIMESTAMP_NANOS_NTZ:
+            return Type.TIMESTAMP_NANOS_NTZ;
+          case UUID:
+            return Type.UUID;
+          default:
+            throw new UnknownVariantTypeException(typeInfo);
+        }
+    }
+  }
+
+  /**
+   * Computes the actual size (in bytes) of the Variant value at `value[pos...]`.
+   * `value.length - pos` is an upper bound of the size, but the actual size may be smaller.
+   * @param value The Variant value
+   * @param pos The starting index of the Variant value
+   * @return The actual size of the Variant value
+   * @throws MalformedVariantException if the Variant is malformed
+   */
+  public static int valueSize(byte[] value, int pos) {
+    checkIndex(pos, value.length);
+    int basicType = value[pos] & BASIC_TYPE_MASK;
+    int typeInfo = (value[pos] >> BASIC_TYPE_BITS) & PRIMITIVE_TYPE_MASK;
+    switch (basicType) {
+      case SHORT_STR:
+        return 1 + typeInfo;
+      case OBJECT:
+        return handleObject(
+            value,
+            pos,
+            (info) -> info.dataStart
+                - pos
```

Comment:
I pointed out earlier that adding `pos` to the offsets may be confusing. I think this is a good example, where in order to calculate the size of the object this has to account for `pos` being added in.